addons/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Comm.java

```diff
@@ -0,0 +1,1145 @@
+            method.addRequestHeader("User-Agent", "TeleHeater/2.2.3");
+            method.setRequestEntity(new ByteArrayRequestEntity(data));
+
+            try {
+                rCode = client.executeMethod(method);
+            } catch (IOException e) {
+                logger.debug("Failed to send data {}", e);
+            } finally {
+                // Release the connection.
+                method.releaseConnection();
+            }
+            logger.debug("Returncode: {}", rCode);
+            return rCode;
+        }
+    }
+
+    /**
+     * This function does the decoding for a new message from the device
+     *
+     */
+    public String decodeMessage(byte[] encoded) {
+        String retString = null;
+        byte[] decodedB64 = null;
+
+        decodedB64 = Base64.decodeBase64(encoded);
+        try {
+            /* Check whether the length of the decryptData is NOT multiplies of 16 */
+            if ((decodedB64.length & 0xF) != 0) {
+                /* Return the data */
+                retString = new String(decodedB64, device.getCharSet());
+                return retString;
+            }
+            // --- create cipher
+            final Cipher cipher = Cipher.getInstance("AES/ECB/NoPadding");
+            cipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(device.getCryptKeyPriv(), "AES"));
+            byte[] decryptedData = cipher.doFinal(decodedB64);
+            byte[] decryptedDataWOZP = removeZeroPadding(decryptedData);
+            return (new String(decryptedDataWOZP, device.getCharSet()));
+        } catch (UnsupportedEncodingException | GeneralSecurityException e) {
+            logger.debug("Exception on encoding: {}", e);
+            return null;
+        }
+    }
+
+    /**
+     * This function does the encoding for a new message to the device
+     *
+     */
+    public byte[] encodeMessage(String data) {
+        try {
+            // --- create cipher
+            byte[] bdata = data.getBytes(device.getCharSet());
+            final Cipher cipher = Cipher.getInstance("AES/ECB/NoPadding");
+            cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(device.getCryptKeyPriv(), "AES"));
+            int bsize = cipher.getBlockSize();
+            logger.debug("Add Padding, encrypt AES and B64..");
+            byte[] encryptedData = cipher.doFinal(addZeroPadding(bdata, bsize, device.getCharSet()));
+            try {
+                return (Base64.encodeBase64(encryptedData));
+            } catch (IllegalArgumentException e) {
+                logger.error("Base64encoding not possible: {}", e.getMessage());
+            }
+        } catch (UnsupportedEncodingException | GeneralSecurityException e) {
+            logger.debug("Exception on encoding: {}", e);
+        }
+        return null;
+    }
+
+    /**
+     * This function checks the capabilities of a service on the device
+     *
+     */
+    public void initObjects(String service, KM200CommObject parent) {
```

Comment:
The code below this point (so the rest of this file) looks bizarly complex, what is it doing. Is the protocol with the device to low level or are you manually doing part of the JSON data / structure processing? It would help to clearly separate data parsing, data validation and data processing, if possible in different classes but at least in different methods. 

Can you provide me with some JSON files and give me some pointer how they flow through this code here.